genvm/vm.go

```diff
@@ -246,52 +254,92 @@ func (v *VM) addRewards(lctx ApplyContext, ss *core.StagedCache, tx *sql.Tx, fee
 	return nil
 }
 
-func (v *VM) execute(lctx ApplyContext, ss *core.StagedCache, txs []types.RawTx) ([]types.TransactionWithResult, []types.TransactionID, uint64, error) {
+func (v *VM) execute(lctx ApplyContext, ss *core.StagedCache, txs []types.ExecutableTx) ([]types.TransactionWithResult, []types.Transaction, uint64, error) {
 	var (
-		rd      bytes.Reader
-		decoder = scale.NewDecoder(&rd)
-		skipped []types.TransactionID
-		fees    uint64
-		results []types.TransactionWithResult
+		rd          bytes.Reader
+		decoder     = scale.NewDecoder(&rd)
+		fees        uint64
+		ineffective []types.Transaction
+		executed    []types.TransactionWithResult
+		limit       = v.cfg.GasLimit
 	)
 	for i := range txs {
 		txCount.Inc()
 
 		t1 := time.Now()
-		tx := &txs[i]
-		rd.Reset(tx.Raw)
-		header, ctx, args, err := parse(v.logger, v.registry, ss, tx.ID, decoder)
+		tx := txs[i]
+
+		rd.Reset(tx.RawTx().Raw)
+		req := &Request{
+			vm:      v,
+			cache:   ss,
+			raw:     txs[i].RawTx(),
+			decoder: decoder,
+			limit:   limit,
+		}
+
+		header, err := req.Parse()
 		if err != nil {
-			v.logger.With().Warning("skipping transaction. failed to parse", log.Err(err))
-			skipped = append(skipped, tx.ID)
+			v.logger.With().Warning("innefective transaction. failed to parse",
+				tx.RawTx().ID,
+				log.Err(err),
+			)
+			ineffective = append(ineffective, types.Transaction{RawTx: tx.RawTx()})
 			invalidTxCount.Inc()
 			continue
 		}
-		t2 := time.Now()
-		transactionDurationParse.Observe(float64(time.Since(t1)))
+		ctx := req.ctx
+		args := req.args
 
-		v.logger.With().Debug("applying transaction",
-			log.Object("header", header),
-			log.Object("account", &ctx.Account),
-		)
 		if ctx.Account.Balance < ctx.Header.MaxGas*ctx.Header.GasPrice {
-			v.logger.With().Warning("skipping transaction. can't cover gas",
+			v.logger.With().Warning("innefective transaction. can't cover gas and max spend",
+				log.Object("header", header),
+				log.Object("account", &ctx.Account),
+			)
+			ineffective = append(ineffective, types.Transaction{RawTx: tx.RawTx()})
+			invalidTxCount.Inc()
+			continue
+		}
+		if limit < ctx.Header.MaxGas {
+			v.logger.With().Warning("innefective transaction. out of block gas",
+				log.Uint64("block gas limit", v.cfg.GasLimit),
+				log.Object("header", header),
+				log.Object("account", &ctx.Account),
+			)
+			ineffective = append(ineffective, types.Transaction{RawTx: tx.RawTx()})
+			invalidTxCount.Inc()
+			continue
+		}
+
+		// NOTE this part is executed only for transactions that weren't verified
+		// when accepted into mempool
+		if !tx.Verified() && !req.Verify() {
```

Comment:
i don't want vm module to have access to already parsed tx header, someone may want to reuse it and it will cause confusion. this interface on the other hand is optional, in a sense that you may use it to improve perf